src/pg_copy.c

```diff
@@ -0,0 +1,696 @@
+		{
+			appendStringInfoChar(buf, sep);
+			appendStringInfoString(buf, strVal(lfirst(cell)));
+			sep = ',';
+		}
+		appendStringInfoChar(buf, ')');
+	}
+	appendStringInfoString(buf, "FROM STDIN");
+
+	if (copyStatement->options)
+	{
+		appendStringInfoString(buf, " WITH ");
+		sep = '(';
+		foreach(cell, copyStatement->options)
+		{
+			DefElem *def = (DefElem *) lfirst(cell);
+			appendStringInfo(buf, "%c%s '%s'", sep, def->defname, defGetString(def));
+			sep = ',';
+		}
+		appendStringInfoChar(buf, ')');
+	}
+	return buf->data;
+}
+
+
+static bool
+PgCopyPrepareTransaction(ShardId shardId, PGconn *conn, void *arg, bool status)
+{
+	PgShardTransactionManager const *tmgr =
+		&PgShardTransManagerImpl[PgShardCurrTransManager];
+	PQputCopyEnd(conn, NULL);
+	return tmgr->Prepare(conn);
+}
+
+
+static bool
+PgCopyAbortTransaction(ShardId shardId, PGconn *conn, void *arg, bool status)
+{
+	PgShardTransactionManager const *tmgr =
+		&PgShardTransManagerImpl[PgShardCurrTransManager];
+
+	if (status)
+	{
+		tmgr->RollbackPrepared(conn);
+	}
+	else
+	{
+		PQputCopyEnd(conn, "Aborted because of failure on some shard");
+		tmgr->Rollback(conn);
+	}
+	PQfinish(conn);
+	return true;
+}
+
+
+static bool
+PgCopyEndTransaction(ShardId shardId, PGconn *conn, void *arg, bool status)
+{
+	PgShardTransactionManager const *tmgr =
+		&PgShardTransManagerImpl[PgShardCurrTransManager];
+
+	Assert(status);
+	tmgr->CommitPrepared(conn);
+
+	PQfinish(conn);
+	return true;
+}
+
+
+static void
+InitializeShardConnections(CopyStmt *copyStatement,
+						   ShardConnections *shardConnections,
+						   ShardId shardId,
+						   PgShardTransactionManager const *transactionManager)
+{
+	ListCell *taskPlacementCell = NULL;
+	List *finalizedPlacementList = LoadFinalizedShardPlacementList(shardId);
+	int placementCount = list_length(finalizedPlacementList);
+	List *failedPlacementList = NULL;
+	ListCell *failedPlacementCell = NULL;
+
+	shardConnections->placements =
+		(PlacementConnection *) palloc0(sizeof(PlacementConnection) * placementCount);
+	shardConnections->status = (bool *) palloc0(placementCount * sizeof(bool));
+	placementCount = 0;
+
+	foreach(taskPlacementCell, finalizedPlacementList)
+	{
+		ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+		char *nodeName = taskPlacement->nodeName;
+		PGconn *conn = NULL;
+		conn = ConnectToNode(nodeName, taskPlacement->nodePort);
+		if (conn != NULL)
+		{
+			char const *copy = ConstructCopyStatement(copyStatement, shardId);
+			shardConnections->placements[placementCount].conn = conn;
+			shardConnections->placements[placementCount].id = taskPlacement->id;
+			placementCount += 1;
+
+			/*
+			 * New connection: start transaction with copy command on it.
+			 * Append shard id to table name.
+			 */
+			if (!transactionManager->Begin(conn) ||
+				!PgShardExecute(conn, PGRES_COPY_IN, copy))
+			{
+				failedPlacementList = lappend(failedPlacementList, taskPlacement);
```

Comment:
We should probably distinguish between Begin or PgShardExecute failing, and connection failing. Begin could fail due to some issue with the DTM. PgShardExecute could fail because of some issue with the worker node (e.g. shard was deleted). I'm not sure if marking the placements as inactive is the way to go there. Then again, the inactive placements only come in if some of the placements fail. Anyway, let's discuss this in the meeting :)